### ISSUE TITLE
Optimizing TIA rendering

### ISSUE THREAD

**Issue by thrust26** [stella-emu/stella#185]
The code seems to be optimizable. Simply moving the vblank() check to the beginning improves maximum framerate by 20..25% in my tests.

And maybe it would be better to retrieve to colors in priority order and stop retrieving when an object is enabled?

EDIT: Looks like my tests are not reliable. The difference seems much smaller than I first thought (<5%).

BTW: I noticed that at maximum frame rate, the sound gets **way** behind.

**Comment by thrust26** [comment]
Worked on the method and profiled my results with VS with Boulder Dash's attraction mode for 90 seconds. The attached code improves the CPU usage of _renderPixel_ from 7.4% to 5.4% (=~35% faster).

So overall Stella improved by ~2%. 😄 

[tia.zip](https://github.com/stella-emu/stella/files/1202956/tia.zip)

**Comment by sa666666** [comment]
Implemented in https://github.com/stella-emu/stella/commit/18568cfbeb2b1d235bfdabeee99e15a406d4ab2a.  I only made one change; I moved the test for vblank() to the ELSE part instead of the IF part.  My reasoning is that the true part actual rendering) happens much more often than the false part (vblank).  If this regresses your improvements, I can change it back.

**Comment by thrust26** [comment]
I doubt it makes any (even measurable) difference.

**Comment by DirtyHairy** [comment]
:+1: Nice that you managed to squeeze something out this way --- `renderPixel` didn't stand out as  a hotspot when I was profiling the code. At which optimization level is VS profiling the code? I would have assumed that, after inlining, the gain of short-circuiting the logic should be barely noticable. The reason why all individual `render` calls happen in every evaluation is mainly historical, I think --- more logic used to be hidden there.

**Comment by thrust26** [comment]
Full Optimization /Ox 
Maybe the VS optimization is less good than your compiler?

The main gain (~80%) is coming from skipping the color calculation during VBlank. Maybe this can be even improved by using YStart and Height?

**Comment by sa666666** [comment]
I'm surprised that TIASurface::render() is happening so quickly (not complaining though).  That's where the image is scaled, TV effects performed, etc.  Did you try testing on a ROM that uses phosphor mode, and also have TV effects turned on?  I'll bet in such a case, that method will take more time.

**Comment by thrust26** [comment]
With TV effects enabled TIASurface::render and AtariNTSC::render really come into play. 
![stella_profiling_rgb](https://user-images.githubusercontent.com/24360774/29034254-ff38a9b8-7b97-11e7-95ea-b25c76f37201.png)

**Comment by sa666666** [comment]
OK, can you now use TV effects but _not_ phosphor?  I suspect that AtariNTSC::render will be the same, but TIASurface::render will be much faster.

**Comment by thrust26** [comment]
You suspected right. 😃 So phosphor rendering is a spot for optimizing.

Could we optimize height? So that only the really displayed area is rendered?
![stella_profiling_rgb_only](https://user-images.githubusercontent.com/24360774/29034821-ce9dd43e-7b99-11e7-9b14-2d3cc21c40c9.png)

**Comment by sa666666** [comment]
It's fairly easy to see why this is the case.  Blargg TV effect calls AtariNTSC::render, and Blargg + phosphor calls AtariNTSC::render + then a loop to copy the old buffer to the new.  But since we do it after Blargg, the framebuffer is 560 pixels wide instead of 160.  You may recall we all discussed this in some detail in another thread.  Doing the phosphor before Blargg would result in it happening on a smaller image, but then Blargg would have to modified to work with phosphor mode (and we determined that to be too difficult.

The code is at lines 379-397 of src/emucore/TIASurface.cxx.  I don't know if we can optimize this further.

BTW 1:  The code currently only renders based on the height that is seen.  So no savings there.  And it only renders the smallest possible image, and leaves stretching to the hardware.  So no savings there either (it will process an image of ~160x200 for non-Blargg, and 560x200 for Blargg).

BTW 2:  If we're going to continue this discussion here, perhaps it's best to reopen this issue and name it something else :smile:

**Comment by thrust26** [comment]
> The code currently only renders based on the height that is seen. 
Yes and no. myTIA->height() returns the height as defined in Game Properties or 210 if nothing is defined. Depending on the game, I am sure this can be reduced. Either by allowing values below 210 and/or by detecting when the last object is displayed (e.g. VBLank is enabled). I remember we discussed this before an it turned out to be a bit more complex.

As for accelerating the code, I have two ideas for now:
1. The values returned by getPhosphor should be precalculated (256x256 bytes + the same for grayscale). 18.2% of the 19.9% are spend in getRGBPhosphor.
2. Maybe cache the getRGBPhosphor values? Since 2^24x2^24 is way too big, maybe we could use a hashmap instead? It depends on how many different values can occur. Probably too many.

So first we should try 1.

**Comment by sa666666** [comment]
I will try (1) in the next day or so.  We can deal with the height later, if we can deal with the complexities you mentioned.

**Comment by thrust26** [comment]
Already implemented (1) with nice results (overall: 19.7% -> 11.2%, getRGBPhosphor(): 18.2%->8.1%):
[TIASurface.zip](https://github.com/stella-emu/stella/files/1205964/TIASurface.zip)
![stella_profiling_rgb_phosphor_new](https://user-images.githubusercontent.com/24360774/29043039-c9a12402-7bb9-11e7-9a03-7921ab1e2db0.png)

**Comment by thrust26** [comment]
> EDIT: I'm not sure how to interpret the data; what is the difference between "Total CPU" and "Self CPU"? Also, I don't know if I even use full optimization when making Stella builds for WIndows. Perhaps I should start doing that??

AFAIK  _Total CPU_ returns the aggregated time of the routine and all subroutines, while _Self CPU_ only shows the time spend in the routine itself (and inlined code). So the sum for column _Self CPU_ should be 100%.

Not sure if full optimization really makes a difference.

**Comment by thrust26** [comment]
The parameter _shift_ in _getRGBPhosphor()_ seems to be only used for the debugger.  Isn't phosphor disabled for the debugger?

**Comment by sa666666** [comment]
Yes, the shift only seems to be used in the debugger.  Phosphor for the debugger isn't disabled, but since you can only see one frame at a time, it effectively _is_ disabled.  So I think we can remove the shift parameter from ```getRGBPhosphor``` completely.

I added your code in https://github.com/stella-emu/stella/commit/1dc78a9b5b8bb809e59f8a24e630d812609be373.  I made the phosphor LUT be calculated just once, in the c'tor.  And I removed the shift parameter as discussed above.  Please retest to confirm the speedups.

**Comment by thrust26** [comment]
I think you have to move the phosphor calculation back, because blend value changes have no effect now until you reload.

**Comment by sa666666** [comment]
Oops, fixed in https://github.com/stella-emu/stella/commit/8759f340e874df2f4cdc9b05192ce3eaa366ca00.  I looked at it and thought "we're going through the same loop, so why do it every time", but I forgot that ```getPhosphor()``` is dynamic, and depends on the current blend level.

**Comment by thrust26** [comment]
Profiling with VS doesn't work at the moment, it's only outputting addresses, not methods. I had to stash my local changes, probably that changed some settings. May take a while until I got that fixed.

**Comment by sa666666** [comment]
OK, there are two more possibilities to consider:
- as @thrust26 mentioned, using the actual height of the image (ie, figuring out exactly where, height-wise, the display starts and stops, and only rendering there (and blanking the rest)
- for phosphor mode, doing the blending before Blargg

The first will require help from the TIA, as only it can know where it starts and stops.  So we need to speak to @DirtyHairy about that one.

The second could potentially be faster than now, since pre-Blargg the image is 160 pixels wide, and post-Blargg it's 560 pixels wide.  That's a decrease of 3.5x if we do blending before Blargg.  What remains to be seen is (a) is this semantically correct (ie, will it still lead to correct output) and (b) whether that new approach/code will be fast enough to not negate any of the savings.

**Comment by sa666666** [comment]
I don't know if you want to try this, but consider having a look at https://tortoisegit.org/docs/tortoisegit/tgit-dug-reset.html.  It details how to do a hard reset on your local repo, which basically resets everything back to the way it was at initial checkout.  Maybe it's better than using stash??

**Comment by thrust26** [comment]
Regarding the 2nd point, I still think phosphor has to be done after Blargg. That's because Blargg mostly emulates  the effects of the electronics behind the screen, while phosphor is happening directly on the screen. Thus it happens later in the picture generation.

However, if we are really brave, we could try to integrate phosphor into Blargg and **maybe** save some time this way.

**Comment by sa666666** [comment]
I am content to not have to touch Blargg if we can get our speedups in another way.  We already discussed that one in some detail, and as you say, it seems that it _should_ be done in the current order anyway.

**Comment by thrust26** [comment]
I started working at a new project this week. There they are using GIT, so I have to learn it now anyway. 😃

**Comment by sa666666** [comment]
You can't escape GIT :smiling_imp:

**Comment by thrust26** [comment]
🏃

**Comment by thrust26** [comment]
OK, profiling works again. And the results are as expected.

**Comment by sa666666** [comment]
OK, I'm logging off for the evening.  Maybe you can try my suggestion from above (on line 388 in TIASurface.cxx, change ```AtariNTSC::outWidth(kTIAW)``` to ```kTIAW``` and re-test).  The image will still be displayed, but phosphor will only be applied for 160 pixels on each line.  But it will tell use the maximum possible speedup we could get.

**Comment by thrust26** [comment]
Maximum speed should be about 8.1%/3.5. So we would gain 5-6%.

**Comment by DirtyHairy** [comment]
Git will find you, whereeer you hide :smile:

**Comment by DirtyHairy** [comment]
The profiles look pretty similar to my old findings with GCC and cachegrind in #7. I'll redo and make screenshots later this week so we can compare the two compilers.

> The main gain (~80%) is coming from skipping the color calculation during VBlank. Maybe this can be even improved by using YStart and Height?

`renderPixel` is only called if the `FrameManager` is in rendering state, so this information is already being used :smirk: In fact, I am surprised that this change is that effective --- I'll have to port it to 6502.ts I guess.

**Comment by thrust26** [comment]
> renderPixel is only called if the FrameManager is in rendering state, so this information is already being used 😏 In fact, I am surprised that this change is that effective --- I'll have to port it to 6502.ts I guess.

Hm, then something in the code seems not quite right, doesn't it?

**Comment by DirtyHairy** [comment]
> Hm, then something in the code seems not quite right, doesn't it?

That's possible, but if `FrameManager::isRendering` were broken, the code should have segfaulted long ago :smirk:

**Comment by thrust26** [comment]
I meant (if I got you right), then renderPixel should not be called while VBlank is active, but yet it is. So there might be room for improvement?

**Comment by DirtyHairy** [comment]
I'll definitely have a closer look, it still may be a smoking gun of some other defect.

**Comment by thrust26** [comment]
I debugged renderPixel() again. vblank() never returns true, so the else part is never executed. 😕 ❓ 

Whatever caused my improvements, its not the vblank() check. Maybe the compiler can optimize this better?

**Comment by sa666666** [comment]
Maybe it's not the vblank() check at all.  You also changed methods, and introduced IF statements.  Previously, all 7 methods were always called, but now there's the possibility to short-circuit and not do all of them.  Perhaps this is where the improvement comes from.

**Comment by DirtyHairy** [comment]
> Could this be due to the different heights and positions of the various kernels (copyright notice, AtariAge logo, title screen and game screen)?

Definitely, that's what I meant before.

Have you checked how stable the numbers for your performance improvements (and their attribution to the various changes you made) are between different profiler runs?

**Comment by thrust26** [comment]
Pretty stable, though by far not as precise as indicated by the two digits after the comma. The timing varies by less than 1% (up to ~0.5%). 

But it would be cool if you could verify some results.

**Comment by thrust26** [comment]
More ideas for improvements:
- @DirtyHairy z26 used to identify areas, which have been changed during the last frame and update only those. Maybe we could do something similar for Stella?
- @sa666666 We could use the GPU for TV effects and phosphor. Does SDL2 (or a different library) provide direct hardware rendered access?
- Use multiple threads for rendering. Even slow hardware often has multiple cores.

**Comment by sa666666** [comment]
Responding to each point in turn:
- What you're talking about is called 'dirty rectangle' updates.  Stella is using SDL2 behind the scenes, and in most cases it is using hardware acceleration (Direct3D for Windows, OpenGL for Linux, OSX).  Conventional wisdom is that dirty rects don't work for hardware acceleration; it's often faster to send the entire image (160x200 or 560x200) that to track where the changes are and only send _that_ data.
- This is definitely an option, but greatly increases complexity.  SDL2 definitely has the ability to drop back to lower levels; in fact, up to release 3.9.3 (last in the 3.x series), it was using OpenGL entirely.  I wrote hand-crafted code for rendering, but still used software for doing TV effects and phosphor.  The main problem is that once we go to a specific toolkit, we're stuck with it.  So that means OpenGL, since it's the most cross-platform.  And then we create shader programs (or similar) that are OpenGL-only.  Problem is that OpenGL isn't everywhere, and Direct3D works much better in Windows.  The current code uses SDL2 abstraction to interface with the best possible hardware libraries on each system.  We break the abstraction by going lower-level, and all the problems that entails.
- I would only want to go to threading as an absolute last resort; IMO the added complexity isn't worth it for what we'd gain.

Only updating the image from start scanline to end scanline is a good compromise.  The entire data itself would still be sent to the video card, but we'd save on calls to getPhoshor and friends for lines that don't need it.  After all, sending the data to the card only accounted for 2% runtime; it's the phosphor calculations that kill it.

I think we may have to accept that, barring any further optimizations like those from the past week, the new core is just slower than the old one.  Not a big surprise really, as it's emulating at a much lower and more accurate level.

And this will continue when we get to the sound code.  It looks like it will be slower too, but again, much more accurate.

**Comment by thrust26** [comment]
> Conventional wisdom is that dirty rects don't work for hardware acceleration

True, but we are doing software rendering for TV effects and phosphor, aren't we? There dirty rectangles might work, no?

> I would only want to go to threading as an absolute last resort; IMO the added complexity isn't worth it for what we'd gain.

I only wanted to use it for the loops in TIASurface::render(). My threading knowledge is **very** limited, is it really that complex?

**Comment by thrust26** [comment]
Minimal changes to getRGBPhosphor. Profiling indicates a **small** speedup. 

```
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
inline uInt32 TIASurface::getRGBPhosphor(const uInt32 c, const uInt32 p) const
{
  #define TO_RGB(color, red, green, blue) \
    const uInt8 red = color >> 16; const uInt8 green = color >> 8; const uInt8 blue = color;

  TO_RGB(c, rc, gc, bc);
  TO_RGB(p, rp, gp, bp);

  // Mix current calculated frame with previous displayed frame
  const uInt8 rn = myPhosphorPalette[rc][rp];
  const uInt8 gn = myPhosphorPalette[gc][gp];
  const uInt8 bn = myPhosphorPalette[bc][bp];

  return (rn << 16) | (gn << 8) | bn;
}
```

**Comment by sa666666** [comment]
Yes, I guess dirty rects might work for tracking changes in the software rendering part of the pipeline.  I guess only drawing from start scanline to end scanline is a form of this.  The (potential) issue is that tracking what has changed is just as slow as just doing the normal work.  But you're welcome to give it a try :smiling_imp: 

IME, threading itself conceptually isn't that difficult.  Using threading and actually getting a speedup is much harder.  Sometimes you do just as much work with thread synchronization as you would by not using threading at all.  Sort of like what I describe in the previous paragraph, I guess.  And then there are times that threading won't help because the process is inherently non-parallel.

**Comment by thrust26** [comment]
I suppose only @DirtyHairy can evaluate the overhead for dirty rectangles without implementing it. Let's wait what he says.

I remember those threading problems, but the task here seems simple enough. And the execution time seems to be constant for each line, so it should be rather easy to get them running in parallel. Or am I missing something here? We could give it a try for phosphor only.

**Comment by sa666666** [comment]
For the loops in TIASurface::render(), it _might_ also be beneficial to use a pointer directly, instead of accessing it with [] and addition.  But we're really getting into micro-optimizations at this point.

**Comment by thrust26** [comment]
One more idea:
- We could save some shifting if we use three different myPhosphorPalettes. But the memory overhead is probably not worth it.

**Comment by sa666666** [comment]
Considering that each palette is 64KB, yes, probably not worth the overhead.

**Comment by sa666666** [comment]
Threading wouldn't work on Blargg anyway, I think, since each pixel is formed by looking at its nearest neighbours (both above and below the current scanline), so it necessarily can't do one line until the previous one is done.

**Comment by sa666666** [comment]
Implemented the change mentioned above in https://github.com/stella-emu/stella/commit/1e7e4cbe207f1e10dc851f07204735897eb2eab4.

**Comment by thrust26** [comment]
I have to look at the code again, but I think Blargg works in independent 7 horizontal pixel chunks.

Anyway, I mainly thought about phosphor.

**Comment by DirtyHairy** [comment]
**Dirty Rectangles**
I think the only way to do this is to compare framebuffers at the pixel level as we are rendering them; anything else is more or less impossible. The new core is a simulation of the TIA at the counter level and lacks any concept of the general image composition. I think this is the reason why it is so precise: it doesn't try to establish any concept of geometrical object (and copies), it is just dumb counters.

If anyone comes up with a smart algorithm to identify dirty areas by examining the framebuffer at the pixel level, but I don't think the complexitiy is worth the speedup.

**Threading**
I think the main danger with threading is the potential of data races. However, spreading the phosphor calculation over multiple cores shouldn't be too difficult. I would consider OpenMP for doing this; it takes away the overhead of spawning, pooling and managing threads. Basically, you just add pragmas to the loop that you want to spread over multiple cores and to the memory that is shared. I have used it in the past with FORTRAN, and it is pretty easy to add for existing code.

**Comment by sa666666** [comment]
The TIA class originally had two framebuffers, current and previous.  They were used to implement phosphor in the old way (static palette only), but also in older versions to do exactly as @DirtyHairy says; track changes at the pixel level.  Basically, each time a pixel was updated, an IF was done to check if something changed.  It was eventually removed, and in general I can't see how checking every single pixel can somehow make anything faster.  As soon as you've checked a pixel, you've done the work.  The whole point was to not look at it in the first place :smiley: 

I am of two minds on the move from OpenGL in Stella 3.x to the SDL2 abstractions in Stella 4.x and above.  On one hand, it greatly decreased my workload and means the same code is used on all systems, with SDL itself picking the best possible toolkit for each platform.

On the other hand, it means not being able to take direct advantage of cool features of modern video card (shaders, hugely parallel hardware, etc), and how the algorithms would run much faster on such cards.  I just wish there was one standard we could use.  Windows is Direct3D only, and does everything it can to break other toolkits.  OpenGL is mostly everywhere, but is also mostly substandard in Windows.  And with Vulkan coming along, I'm not sure how long OpenGL will be around.

Many other emulators are taking the OpenGL-only approach and greatly benefiting from using OpenGL directly.  But one of the design goals of Stella is to be able to run anywhere.  Arrgh, I've had this conversation with myself many times over the years, and never really find a best solution.

**Comment by thrust26** [comment]
> Arrgh, I've had this conversation with myself many times over the years, and never really find a best solution.

Blame Microsoft!

**Comment by sa666666** [comment]
Oh, I do, with the burning hatred of a thousand suns :smiling_imp: But at the end of the day, I'm still back where I started, and still have to figure out the best way forward.

**Comment by thrust26** [comment]
Just for fun, I tried to implement multi threading for phosphor calculation. Code by Google. 😄 
[TIASurface_Multi.zip](https://github.com/stella-emu/stella/files/1208739/TIASurface_Multi.zip)

**Comment by DirtyHairy** [comment]
Again, I'd propose to use [OpenMP](https://en.wikipedia.org/wiki/OpenMP): it will take care of spawning and pooling threads, about memory sharing and will automatically select an appropiate number of threads to use in order to ensure optimal use of the hardware. Also, you don't need any boilerplate apart from adding pragmas to the loop that should be optimized.

**Comment by thrust26** [comment]
No idea, I don't know how to profile that. But CPU went well above 30% on my quad core at maximum framerate,

**Comment by thrust26** [comment]
VS only supports OpenMP 2.5. 😞

**Comment by DirtyHairy** [comment]
I *think* 2.0 should be enough 😉 I can try a sample implementation later (propably not today though).

**Comment by sa666666** [comment]
@thrust26, but what does the CPU do at maximum framerate _before_ you added this code?  The reason I ask is that sometimes threading is deceiving, and adding it makes things slower.  I haven't tested or really looked in any detail, but I notice that the thread is using a lambda with '=' capture specification, meaning to copy everything.  Now it may be required to do that, but it's also a possible speed hit.

As for OpenMP vs. rolling our own, there's a certain appeal to using C++11 threads directly.  The biggest being that we don't add another dependency to the codebase (OpenMP).  But I really don't have much experience in this area either way, so I will wait for your feedback.

**Comment by thrust26** [comment]
@sa666666 I can't see any difference. Is there any way to display the framerate?

@DirtyHairy Is it reallly that easy? Use _/openmp_ option and put `#pragma omp parallel for` above the loop?

**Comment by thrust26** [comment]
> As for OpenMP vs. rolling our own, there's a certain appeal to using C++11 threads directly

Agreed.

**Comment by sa666666** [comment]
@thrust26, no there's no way to know the actual framerate at this time.

However, now that I think about it, threading isn't going to make things faster anyway, at least in terms of CPU usage.  Inefficiencies aside, what we can do in one loop and 1 thread should be the same as 1/4 of the work each in 4 threads.

I guess this is more beneficial for lower-clocked systems, where it can't do all the work in the 16.7ms per frame and as a result get frames skipped.  But overall, the work is still being done, just spread 'sideways'.  This is different from our previous optimizations, which resulted in code that actually did less work.

**Comment by sa666666** [comment]
I also want to add that threading has come a long way since I last looked at it.  OpenMP seems dead-easy, but even the C++11 version is very simple compared to the last time I used it (C and pthreads).  So my apparent disdain for threading is completely out of date.

**Comment by thrust26** [comment]
For sure threading doesn't improve code performance. It just reduces execution time.

Do you have an overview on what platforms Stella is used? Especially at the lower end.

**Comment by sa666666** [comment]
The 'big 3' (Linux, OSX, and Windows) on commodity hardware probably make up 95% or more of usage.  As with most apps, Windows probably is the majority (I would estimate > 80%), and OSX and Linux the remaining.  So on those systems, none of these optimizations are really necessary anyway.  But IMO, we should always strive for maximum efficiency.

However, there are many emulation boxes that run Linux on lesser hardware.  These tend to be ARM based, but I don't know if they have multiple cores.  Probably most of the current ones do.  The RasPI is a very common one.  I've also been contacted by one of those 'emulation box' companies about using Stella is a fully-licenced GPL-approved way (more discussion by PM if you're interested).  On that particular system, they note that their proprietary 250Mhz system is too slow, but a 40Mhz RasPI is sufficient.  Of course this is the general emulation itself I'm talking about here, so that's without phosphor or TV effects at all.

I think getting full emulation, TV effects and phosphor mode (and the new sound core) running on a 400Mhz or above RasPI would be a good thing to aim for.

**Comment by thrust26** [comment]
Which RasPi model? 

And yes, I am interested into that 'emulation box'. 😄

**Comment by sa666666** [comment]
I suspect most people are trying it on the latest model; so version 3b.  I guess this would be the one to aim for.  But since it's running at 1.2GHz, I think it's not a problem anyway.  The really low-end ones are the ones used in the emulation boxes, which is generally why they have crappy emulation.

Anyway, I have to step away for a few hours; have to get some work done.

**Comment by DirtyHairy** [comment]
@thrust26 it's been some 5 years since I last used OpenMP, but basically, yes 😉

**Comment by thrust26** [comment]
I optimized the phosphor loops by unrolling (2x and 8x, ~130 bytes each) and smoothing the code. TIASurface::render() went from ~11.2% to ~9.0% in profiling. 

[TIASurface_unrolled.zip](https://github.com/stella-emu/stella/files/1209290/TIASurface_unrolled.zip)

**Comment by DirtyHairy** [comment]
@thrust26 I think what is missing for OpenMP is pragmas to define the memory semantics. Iirc, the default is to give each thread its own copy of all quantities, that's why the lower half of the screen is not updated.

**Comment by sa666666** [comment]
@thrust26, it's unclear to me why you use 2 and 8; what is the significance of those magic numbers?  Or more to the point, why they aren't the same (since the two blocks of phosphor code are essentially the same)??

**Comment by thrust26** [comment]
I changed the code into this. It works better than before, but now there are distortions. Any suggestion? 
```
      Int32 bufofs;
      // Note: The codes assumes that AtariNTSC::outWidth(kTIAW) == outPitch
#pragma omp parallel for
      for (Int32 y = 0; y < height; ++y)
      {
        bufofs = y * AtariNTSC::outWidth(kTIAW); 
        for(Int32 x = AtariNTSC::outWidth(kTIAW) / 8; x > 0; --x)
        {
          // Store into out buffer and back into displayed frame buffer (for next frame)
          rgbIn[bufofs++] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          rgbIn[bufofs++] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          rgbIn[bufofs++] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          rgbIn[bufofs++] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          rgbIn[bufofs++] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          rgbIn[bufofs++] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          rgbIn[bufofs++] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          rgbIn[bufofs++] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
        }        
      }
```

**Comment by thrust26** [comment]
> @thrust26, it's unclear to me why you use 2 and 8

I found that more than 8 times unrolling doesn't give any relevant improvement. And since Blargg lines are 3.5 times wider than non-Blargg lines, I thought less unrolling (8/3.5 = ~2) would be OK. But you can unroll as often as you want.

**Comment by sa666666** [comment]
Isn't it possible that this loop unrollling will start to interfere with threading?  Maybe we should test both methods before doing any further commits.

@thrust26, so the 2 and 8 are essentially from empirical evidence?

**Comment by sa666666** [comment]
@thrust26, each line now depends on the other one.  In particular, the value for bufofs in one line is incremented and then used in the second line.  So there is now dependency between the lines.

Also, on my compiler at least, your code (without threading) is complaining about sequence points.  I think that's related to doing an increment and using the bufofs variable in the same line.

**Comment by thrust26** [comment]
@sa666666 Yes, 8 is empirical (2 is only deducted from 8).

As far as I can tell, the loop is not split **inside**. So the unrolling is fine. But the the inner loop itself is split too. That caused the problems. Fixed above.

**Comment by sa666666** [comment]
@thrust26, not related to threading, but you need to split each line from
```
rgbIn[bufofs++] = out[pos++] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
```
to
```
rgbIn[bufofs] = out[pos] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
bufofs++;  pos++;
```
Otherwise, you're actually violating a rule of the language.  In particular, it's not guaranteed that result of '++' is done before or after it is used later in the line.  I don't know how much, if any, this will affect the performance.

**Comment by thrust26** [comment]
I wondered about that ++ too. It works, but I will better change that.

Meanwhile I solved my problem. Here is the code which only threads the outer loop and causes no problem (bufofs has to be local inside the outer loop). I think this is more effective than doing the math in the inner loop. But I can't measure it.
```
      // Note: The codes assumes that AtariNTSC::outWidth(kTIAW) == outPitch
#pragma omp parallel for
      for (Int32 y = 0; y < height; ++y)
      {
        Int32 bufofs = y * AtariNTSC::outWidth(kTIAW);
        for (Int32 x = AtariNTSC::outWidth(kTIAW) / 8; x > 0; --x)
        {
          // Store into out buffer and back into displayed frame buffer (for next frame)
          rgbIn[bufofs] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          bufofs++;
          rgbIn[bufofs] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          bufofs++;
          rgbIn[bufofs] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          bufofs++;
          rgbIn[bufofs] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          bufofs++;
          rgbIn[bufofs] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          bufofs++;
          rgbIn[bufofs] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          bufofs++;
          rgbIn[bufofs] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          bufofs++;
          rgbIn[bufofs] = out[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
          bufofs++;
        }
      }
```

**Comment by thrust26** [comment]
Ouch! When using threading, the CPU utilization goes to ~70%!!! 

Any idea what's happening? Too much overhead from OpenMP? Or maybe cache problems?

**Comment by sa666666** [comment]
The '++' only _works_ in that Visual Studio doesn't complain about it, and it happens to work.  But it is against the spec.  I typically compile in Visual Studio, gcc and clang when testing new features, and there are quite a few times that one compiler picks up something that the other doesn't.

As for testing the efficacy of the new approach, let's see what @DirtyHairy has to say.  He seems to have more experience with OpenMP and threading in general, so maybe he can suggest a possible test.

**Comment by sa666666** [comment]
Yep, that's what I was afraid of with threads.  We sit down and hammer out an approach.  Everything seems to fine, make logic sense, etc.  And then test it, and it ends up being slower :disappointed:

**Comment by thrust26** [comment]
True. But my hand written code works nicely. So we are fighting the frame work here, not the threading in general.

**Comment by sa666666** [comment]
When you say "hand written code", I assume you mean the version using C++11 threads directly?

Also, again I'm not sure about exactly how threading works, but isn't it the same as calling a bunch of functions?  So maybe there's the function call overhead?? :confused:

**Comment by thrust26** [comment]
Yup, my Google code. :smiley:

Yes, there is function overhead. And some math overhead in our example too. So the total CPU time will increase. But at least in my own, handwritten code, the overhead is minimal.

BTW: Having a chat channel as @DirtyHairy suggested would be really nice.

**Comment by thrust26** [comment]
I just tried to tell OpenMP how many threads to use: 1 thread = 6%, 2 threads = 30%. So CPU utilization already jumps with just one extra thread.

@DirtyHairy mentioned a chat in the trackball thread:
> As a sidenote, @sa666666 , I've never tried it before, but [gitter](https://gitter.im/) is a chat that is connected to github projects and might be more useful than tickets for discussions like this :smirk: I'm tempted to do an experiment and set it up of Stella, if this is OK for you.

**Comment by DirtyHairy** [comment]
@sa666666 You can set up gitter on the [web page](https://gitter.im/). I just tried to do so, but I would need administrative access for the stella-emu organization :smirk:

Concerning the threading: in my experience, if used correctly, the overhead from OpenMP is negligible. It mainly consists of the bookkeeping for spawning and pooling threads and for distributing workloads. I have used OpenMP for numbercrunching before, and the 70% CPU are certainly not normal. What I'd like to do is to try my own implementation of OpenMP in the phosphor code --- not because I dislike Thomas', but I want to look at the code with fresh eyes :smirk: I'll see whether I can come up with something in the next hour.

**Comment by thrust26** [comment]
I want to know before I go to bed. 😄

**Comment by thrust26** [comment]
@sa666666 16 cores? Or 8 cores with Hyper-Threading? i7?

**Comment by sa666666** [comment]
I just created the gitter thing and invited both of you.  I'm not sure how to use it yet, or if it's set up correctly.  Also not sure how it improves on discussing things here, but we will see ...

@thrust26, Ryzen CPU, 8 courses + hyperthreading = 16 total.  Stella compiles in 25 seconds :grinning: 

I'd also like to know what's going on wrt threading.  But I'm soon logging off too.  It's almost 7PM here, and I've been working on Stella today longer than my actual job :smile:

**Comment by thrust26** [comment]
> Otherwise, you're actually violating a rule of the language. In particular, it's not guaranteed that result of '++' is done before or after it is used later in the line. 

Would that be OK or is it still undefined? I would assume that the expression is always evaluated first, before the result is assigned. And the code looks nicer that way. :smiley:

`rgbIn[bufofs++] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);`

**Comment by DirtyHairy** [comment]
That's what my code looks like:
```
    case Filter::Phosphor:
    {
      uInt8*  tiaIn = myTIA->frameBuffer();
      uInt32* rgbIn = myRGBFramebuffer;

      #pragma omp parallel for
      for(uInt32 y = 0; y < height; ++y)
      {
        const uInt32 offset = y * width, offsetOut = y * outPitch;

        for(uInt32 x = 0; x < width; ++x)
        {
          const uInt32 bufofs = offset + x;
          const uInt8 c = tiaIn[bufofs];
          const uInt32 retVal = getRGBPhosphor(myPalette[c], rgbIn[bufofs]);

          // Store back into displayed frame buffer (for next frame)
          rgbIn[bufofs] = out[offsetOut + x] = retVal;
        }
      }
      break;
    }
```
Unfortunately, I see the same as @thrust26 : there is a constant load of 20% per core if OMP is active., even if I completely gut the loop to its bare skeleton. I must admit that I have no idea what is going on here, I have used OMP before without problems :angry:

The only difference to my previous applications was that I was parallelizing applications that never slept and that were trying to crunch number as fast as possible. I *suspect* that the issue is related to Stella sleeping between frames: sleep only puts the main thread to sleep, so it could be that the other threads are caught in a spinlock. I'll do a profile out of curiousity when I find time, but I think that's it for OMP in our usecase :white_flag:

**Comment by DirtyHairy** [comment]
> Would that be OK or is it still undefined?

Afaik it is undefined, but @sa666666 is the authoritative source on that :smirk:

**Comment by thrust26** [comment]
We should not give up that early. Maybe there is a configuration for OpenMP which changes the behavior?

**Comment by thrust26** [comment]
@DirtyHairy Another idea (for TIA class):
In `TIA::renderPixel()` we now check the objects in priority order. This means that low priority objects are not checked when they are not visible. Could this be used to optimize the `tick()` methods?

Maybe, if we reorder e.g. the `tick()` calls like in `TIA::renderPixel()` we could at least save calculating `collision` if the object is hidden. And maybe more?

The savings won't be big (if any), but maybe its worth a look.

**Comment by sa666666** [comment]
> Afaik it is undefined, but @sa666666 is the authoritative source on that :smirk:

I wouldn't say authority, it's just that I make sure to compile everything on 3-4 compilers, to catch any possible issue.  Based on my understanding, the sequence point issue is because you're using variable on both sides of the '=' sign, but on one side it is doing a post-increment.  So saying ```x = x++``` would also complain.

So the easiest way to fix it is:
```
rgbIn[bufofs] = getRGBPhosphor(out[bufofs], rgbIn[bufofs]);
bufofs++
```

**Comment by thrust26** [comment]
OK then.

BTW: Instead of copying into `out[] `one by one, I am using `memcpy()` now. 😄

**Comment by sa666666** [comment]
**My opinions on OMP vs. C++11 threading**:
- If OMP is too difficult or not performant, we should just drop it and attempt C++11 threads
- I'm not entirely convinced that threading will help at all anyway
- We probably shouldn't spend too much time on this

I'm not saying outright to just close the book on this approach, just that we shouldn't be overly consumed by getting it to work. IMO, improvements to the code itself are more important than threading improvements.

@thrust26, what is your current code?  There have been several iterations now, but I haven't committed anything else yet.  Also, maybe you can try your 'google code' with 4 threads instead of 2, to see what kind of performance it gives.

**Comment by thrust26** [comment]
You mean the non-blargg with phosphor code?

**Comment by sa666666** [comment]
@thrust26, what you called 'hand-written' code, or 'google code'.  It's getting hard to keep track of things now, since this issue contains almost 100 messages and several different code examples.

What I was referring to is your original code that uses 2 C++11 threads.  It was operating on the Blargg+phosphor code path, but obviously only applied to the phosphor loop (since we haven't even looked optimizing Blargg yet).

I suggest to use 4 threads instead of 2, to see if (a) it makes things any faster, and more importantly (b) that it doesn't drastically kill performance like OMP.

**Comment by sa666666** [comment]
I feel we're all ping-ponging a bit on this stuff, basically going all over the place.  For example, in the process of optimizing the phosphor loop for use without threading, we may be making it more difficult to do _with_ threading.  So we need to decide which approach to take, and work from there.

Also, it's 10AM here now.  At 1PM (3 hours from now) I won't be available for at least 3 hours (invigilating my final exam), so if there's anything you need from me, now is the time.

**Comment by thrust26** [comment]
There you go! 

See line 40 for enabling passive OMP wait policy. 
`case Filter::BlarggPhosphor:` contains multiple variations of threading (and one without). I have no clue how to measure the effects. Probably a better profiler can do.

[TIASurface_lots_of_threading.zip](https://github.com/stella-emu/stella/files/1211398/TIASurface_lots_of_threading.zip)

**Comment by thrust26** [comment]
Applied threading to Blargg now too. This and cleaned TIASurface attached.

Profiling looks good:
![stella_profiling_multithreaded_with_blarg](https://user-images.githubusercontent.com/24360774/29138219-490b8ea0-7d43-11e7-82ba-db795a698a65.png)

[AtariNTSC.zip](https://github.com/stella-emu/stella/files/1212495/AtariNTSC.zip)

Question: Should we really split into the number of cores? Or should we reserve some cores, e.g. for the OS?

**Comment by sa666666** [comment]
Sorry, still having a brain fart and _still_ not totally understanding the results of those graphs.  In any event, as long as nothing slowed down, I'd be happy :smile: 

As for limiting threads and reserving for the OS, I think 4 is probably enough too.  But I'd like to hear what @DirtyHairy has to say, and also some testing from him in Linux.

**Comment by sa666666** [comment]
@thrust, the following patch solves your issue about passing the phosphor palette to AtariNTSC.  I tried cleaner ways of doing it, but basically resorted to just copying the data.  It's only 64KB, and only done when phosphor changes.
[pass_phosphor.diff.zip](https://github.com/stella-emu/stella/files/1213357/pass_phosphor.diff.zip)

**Comment by thrust26** [comment]
Attached the lastest source code.

1. threads are now limited to 4 max
2. blargg and phosphor are now done in the same thread
[Rendering_Blargg_with_Phosphor.zip](https://github.com/stella-emu/stella/files/1214479/Rendering_Blargg_with_Phosphor.zip)

And here profiling the result twice, this should give an indication about the precision of the values.
![twice_blargg_with_phosphor](https://user-images.githubusercontent.com/24360774/29165736-0c395e5e-7dc4-11e7-85f8-ad5bc6592a0b.png)

**Comment by sa666666** [comment]
The attached git diff patch now compiles in Linux with gcc and clang. @DirtyHairy, you can apply it to your threading branch when you like.
[threading.diff.zip](https://github.com/stella-emu/stella/files/1214989/threading.diff.zip)

**Comment by thrust26** [comment]
An small update to the existing threading (if we want to continue to use it). This now makes the main thread busy too and spawns one extra thread less:
```
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void AtariNTSC::render(const uInt8* atari_in, const uInt32 in_width, const uInt32 in_height,
  void* rgb_out, const uInt32 out_pitch, uInt32* rgb_in)
{
  // spawn myNumThreads - 1 threads...
  for (uInt32 i = 1; i < myNumThreads; i++)
    myThreads[i] = std::thread([=] {
    rgb_in == NULL ? renderThread(atari_in, in_width, in_height, myNumThreads, i, rgb_out, out_pitch) :
      renderWithPhosphorThread(atari_in, in_width, in_height, myNumThreads, i, rgb_in, rgb_out, out_pitch);
  });
  // make the main thread busy too
  rgb_in == NULL ? renderThread(atari_in, in_width, in_height, myNumThreads, 0, rgb_out, out_pitch) :
    renderWithPhosphorThread(atari_in, in_width, in_height, myNumThreads, 0, rgb_in, rgb_out, out_pitch);
  // ...and make them join again
  for (uInt32 i = 1; i < myNumThreads; i++)
    myThreads[i].join();
  
  // Copy phosphor values into out buffer 
  if (rgb_in != NULL)  
    memcpy(rgb_out, rgb_in, in_height * out_pitch);
}
```

**Comment by thrust26** [comment]
Forgot this minor one in `AtariNTSC::initialize()`, line 40: 

`myThreads = new std::thread[myNumThreads - 1];`

**Comment by thrust26** [comment]
Now that the threading code has been committed, shall we close this one or keep it open for more rendering thread alternatives?

**Comment by sa666666** [comment]
I think since we moved discussion of this to gitter, this one can be closed.  As for allowing a choice of threads, is there really much point?  Maybe the choice should simply be on or off??

**Comment by thrust26** [comment]
Agreed, on and off would do for a start.

**Comment by sa666666** [comment]
OK, I'll get it added later this evening.
